### PR TITLE
Distributed Training Support

### PR DESCRIPTION
**Pull request by farhadrgh**
## What's New

- **Context Parallelism**: Split sequences across GPUs
- **Backend Abstraction** (`nvsubquadratic/distributed/backend.py`): Plugin architecture for Megatron/DeviceMesh/FSDP
- **Lightning Strategy** (`nvsubquadratic/strategies/context_parallel_strategy.py`): Extends DDPStrategy with CP support
- **CP-Aware DataModule** (`nvsubquadratic/datamodules/cp_aware_mixin.py`): Zigzag split for data loading
- **Network Integration**: Networks/blocks propagate cp_group to sequence mixers
- **Checkpointing**: Resume training with `--ckpt_path` (handles cluster time limits)

### CODE REVIEW

File: examples/lightning_wrappers.py [hunk 1]
```diff
@@ -180,9 +180,38 @@ def __init__(
...
+
+        Args:
+            x: Input tensor.
+            **kwargs: Additional keyword arguments to pass to network.
+
+        Returns:
+            Output tensor from network.
+        """
+        # Check if network accepts cp_group parameter
+        import inspect
```
Comment (pstjohn) [inline examples/lightning_wrappers.py:205]: maybe don't `import` during the model's forward pass

**Comment by cspades** [comment]
`nvsubquadratic/distributed/backend.py` will work just fine with `DeviceMesh` and Torch DCP (the native Torch version of Megatron's distributed checkpointing that works with DTensor and DeviceMesh).

You have orthogonal mesh dimensions, and it doesn't look like you need to combine any groups either, and there shouldn't be any issue with this API.

Just in case, here's how `MegatronFSDP` handles its DeviceMesh: https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/distributed/fsdp/src/megatron_fsdp/utils.py#L661

The only thing to note is this:
```
        """
        Store a persistent reference to the core device meshes that back Megatron-FSDP.
        This is necessary because _MeshEnv (_mesh_resources) may not persist:
            - _mesh_resources.child_to_root_mapping
            - _mesh_resources.root_to_flatten_mapping
            - _mesh_resources.flatten_name_to_root_dims
            - ...
        during Torch Autograd, so child and flattened sub-meshes may be cleared.
        For example, this breaks Megatron-FSDP when self.dp_shard_dim is the flattened
        sub-mesh of the DP and CP root mesh dimensions.
        FIXME(@cspades): Identify the root cause of this behavior.
        """
```
and I learned today this is a known issue by the Torch team and they are deprecating child mesh access from the root mesh:
```
[rank0]:/usr/local/lib/python3.12/dist-packages/torch/distributed/device_mesh.py:778: UserWarning: Slicing a flattened dim from root mesh will be deprecated in PT 2.11. Users need to bookkeep the flattened mesh directly.
```

File: nvsubquadratic/strategies/context_parallel_strategy.py [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+        if self.backend is not None:
+            dp_group = self.backend.get_data_parallel_group()
+            logger.info("Wrapping model with DDP using CP-aware data parallel group")
+        else:
+            logger.info("Wrapping model with DDP using default process group")
+
+        return DDP(
+            model,
+            device_ids=[self.local_rank] if self.local_rank is not None else None,
+            process_group=dp_group,
```
Comment (jstjohn) [inline nvsubquadratic/strategies/context_parallel_strategy.py:149]: Yeah as long as you are sure that this includes every GPU basically, then you should be ok... The problem could be that when you do CP, some of the GPUs are getting different slices of the same input sample, so from the viewpoint of your dataloaders would be part of the same DP rank. It may be ok, just verify that you are getting all reduces happening across DP+CP. These PRs may be useful: https://gitlab-master.nvidia.com/clara-discovery/boltz/-/merge_requests/97 https://gitlab-master.nvidia.com/clara-discovery/boltz/-/merge_requests/131

Comment (jstjohn) [reply nvsubquadratic/strategies/context_parallel_strategy.py:149]: Looking that over it seems like I did the same thing using the DP group.

Comment (farhadrgh) [reply nvsubquadratic/strategies/context_parallel_strategy.py:149]: Thanks for the feedback @jstjohn. Got this fixed by separating the methods for getting CP+DP and DP-only. Also added `DistributedSampler` and avoided broadcasting CP shards

**Comment by farhadrgh** [comment]
KNOWN ISSUES:
- The classification loss will not work with CP=2 because there is a non-mean aggregation over the
  CP spatial dimension. The loss needs to be mean aggregated over the CP spatial dimension for the
  assumption we make that we can take the average of the gradients over CP ranks in our DDP object.
- GroupNorm will not work with CP=2 the way we call it in the current network implementation
  because it seems to somehow aggregate results over the CP spatial dimension.
- If you address the two above issues by replacing the loss with mean squared activation or
  something similar that has no non-mean aggregation over the CP spatial dimension, as well as
  removing the GroupNorm, then the test will pass with max metrics around 5% and many at 0%.

SelfAttention CP support
- The self-attention implementation has an issue where the number of attention heads is not correctly split across ranks during context parallelism, causing each rank to use the global number of heads instead of a local subset. This means the computation is not properly parallelized and may result in overcomputing, as each rank redundantly processes the full set of heads.
- Additionally, the rope logic does not account for the splitting of channels, which is necessary when the channels are distributed across ranks. This needs to be fixed to ensure correct frequency selection for each head.
- The current tests may not catch this issue because the implementation overcomputes, so the tests pass even though the parallelism is not functioning as intended.
